Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,648 @@
+        map_add' := fun p p' => LinearMap.ext <| fun q => (Submodule.Quotient.eq _).mpr ?_
+        map_smul' := fun r p => LinearMap.ext <| fun q => (Submodule.Quotient.eq _).mpr ?_ }
+  · change q ⊗ₜ[R] (h (p + p')) - (q ⊗ₜ[R] (h p) + q ⊗ₜ[R] (h p')) ∈ range (lTensor Q f)
+    rw [← TensorProduct.tmul_add, ← TensorProduct.tmul_sub]
+    apply le_comap_range_lTensor f
+    rw [← hfg, mem_ker, map_sub, map_add]
+    simp only [hgh _, sub_self]
+  · change q ⊗ₜ[R] (h (r • p)) - r • q ⊗ₜ[R] (h p) ∈ range (lTensor Q f)
+    rw [← TensorProduct.tmul_smul, ← TensorProduct.tmul_sub]
+    apply le_comap_range_lTensor f
+    rw [← hfg, mem_ker, map_sub, map_smul]
+    simp only [hgh _, sub_self]
+
+lemma lTensor.inverse_ofRightInverse_apply
+    {h : P → N} (hgh : Function.RightInverse h g) (y : Q ⊗[R] N) :
+    (lTensor.inverse_ofRightInverse Q hfg hgh) ((lTensor Q g) y) =
+      Submodule.Quotient.mk (p := (LinearMap.range (lTensor Q f))) y := by
+  simp only [← LinearMap.comp_apply, ← Submodule.mkQ_apply]
+  rw [exact_iff] at hfg
+  apply LinearMap.congr_fun
+  apply TensorProduct.ext'
+  intro n q
+  simp only [coe_comp, Function.comp_apply, lTensor_tmul, Submodule.mkQ_apply]
+  simp only [lTensor.inverse_ofRightInverse]
+  simp only [TensorProduct.lift.tmul, coe_mk, AddHom.coe_mk, mk₂_apply]
+  simp only [flip_apply, coe_mk, AddHom.coe_mk, coe_comp, Function.comp_apply, mk_apply,
+    Submodule.mkQ_apply]
+  rw [Submodule.Quotient.eq]
+  rw [← TensorProduct.tmul_sub]
+  apply le_comap_range_lTensor f n
+  rw [← hfg, mem_ker, map_sub, sub_eq_zero, hgh]
+
+lemma lTensor.inverse_ofRightInverse_comp_rTensor
+    {h : P → N} (hgh : Function.RightInverse h g) :
+    (lTensor.inverse_ofRightInverse Q hfg hgh).comp (lTensor Q g) =
+      Submodule.mkQ (p := LinearMap.range (lTensor Q f)) := by
+  rw [LinearMap.ext_iff]
+  intro y
+  simp only [coe_comp, Function.comp_apply, Submodule.mkQ_apply,
+    lTensor.inverse_ofRightInverse_apply]
+
+/-- The inverse map in `lTensor.equiv` -/
+noncomputable
+def lTensor.inverse :
+    Q ⊗[R] P →ₗ[R] Q ⊗[R] N ⧸ LinearMap.range (lTensor Q f) :=
+  lTensor.inverse_ofRightInverse Q hfg
+    (Function.rightInverse_surjInv hg)
+
+lemma lTensor.inverse_apply (y : Q ⊗[R] N) :
+    (lTensor.inverse Q hfg hg) ((lTensor Q g) y) =
+      Submodule.Quotient.mk (p := (LinearMap.range (lTensor Q f))) y := by
+  simp only [lTensor.inverse, lTensor.inverse_ofRightInverse_apply]
+
+lemma lTensor.inverse_comp_rTensor :
+    (lTensor.inverse Q hfg hg).comp (lTensor Q g) =
+      Submodule.mkQ (p := LinearMap.range (lTensor Q f)) := by
+  rw [lTensor.inverse, lTensor.inverse_ofRightInverse_comp_rTensor]
+
+/-- For a surjective `f : N →ₗ[R] P`,
+  the natural equivalence between `Q ⊗ N ⧸ (image of ker f)` to `Q ⊗ P`
+  (computably, given a left inverse) -/
+def lTensor.equiv_ofRightInverse {h : P → N} (hgh : Function.RightInverse h g) :
+    ((Q ⊗[R] N) ⧸ (LinearMap.range (lTensor Q f))) ≃ₗ[R] (Q ⊗[R] P) := {
+  lTensor.toFun Q hfg with
```

Comment:
again I am slightly scared about the `with` but I would defer to Eric or Matt.

Reply:
This file seems pretty fast, but CC @eric-wieser @mattrobball

Reply:
It think it would be fine performance-wise. But, generally I think the code reads better if, when using an instance of the/a parent structure, that you inline it instead of using `with`. `with` is better for when you break direct inheritance but still have all the fields. 

```suggestion
  { toLinearMap := lTensor.toFun Q hfg
```